sql/src/main/java/io/crate/planner/consumer/SemiJoins.java

```diff
@@ -0,0 +1,259 @@
+            return ((QueriedTableRelation) rel).tableRelation();
+        }
+        // TODO: support other cases as well
+        return null;
+    }
+
+    private static void removeRewriteCandidatesFromWhere(QueriedRelation rel, Function rewriteCandidate) {
+        rel.querySpec().where().replace(FuncSymbols.mapNodes(f -> {
+            if (f == rewriteCandidate) {
+                return Literal.BOOLEAN_TRUE;
+            }
+            return f;
+        }));
+    }
+
+    static List<Function> gatherRewriteCandidates(Symbol query) {
+        ArrayList<Function> candidates = new ArrayList<>();
+        RewriteCandidateGatherer.INSTANCE.process(query, candidates);
+        return candidates;
+    }
+
+    @Nullable
+    static SelectSymbol getSubqueryOrNull(Symbol symbol) {
+        symbol = unwrapCast(symbol);
+        if (symbol instanceof SelectSymbol) {
+            return ((SelectSymbol) symbol);
+        }
+        return null;
+    }
+
+    private static Symbol unwrapCast(Symbol symbol) {
+        while (isCastFunction(symbol)) {
+            symbol = ((Function) symbol).arguments().get(0);
+        }
+        return symbol;
+    }
+
+    /**
+     * t1.x IN (select y from t2)  --> SEMI JOIN t1 on t1.x = t2.y
+     */
+    static Symbol makeJoinCondition(Function rewriteCandidate, AnalyzedRelation sourceRel, QueriedRelation subRel) {
+        assert getSubqueryOrNull(rewriteCandidate.arguments().get(1)).relation() == subRel
+            : "subRel argument must match selectSymbol relation";
+
+        rewriteCandidate = RefReplacer.replaceRefs(
+            rewriteCandidate,
+            r -> sourceRel.getField(r.ident().columnIdent(), Operation.READ));
+        String name = rewriteCandidate.info().ident().name();
+        assert name.startsWith(AnyOperator.OPERATOR_PREFIX) : "Can only create a join condition from any_";
+
+        List<Symbol> args = rewriteCandidate.arguments();
+        Symbol firstArg = args.get(0);
+        List<DataType> newArgTypes = ImmutableList.of(firstArg.valueType(), firstArg.valueType());
+
+        FunctionIdent joinCondIdent = new FunctionIdent(AnyOperator.nameToNonAny(name), newArgTypes);
+        return new Function(
+            new FunctionInfo(joinCondIdent, DataTypes.BOOLEAN),
+            ImmutableList.of(firstArg, castIfNeededOrFail(subRel.fields().get(0), firstArg.valueType()))
+        );
+    }
+
+    private static class RewriteCandidateGatherer extends SymbolVisitor<List<Function>, Boolean> {
+
+        static final RewriteCandidateGatherer INSTANCE = new RewriteCandidateGatherer();
+
+        @Override
+        protected Boolean visitSymbol(Symbol symbol, List<Function> context) {
+            return true;
+        }
+
+        @Override
+        public Boolean visitFunction(Function func, List<Function> candidates) {
+            String funcName = func.info().ident().name();
+
+            /* Cannot rewrite a `op ANY subquery` expression into a semi-join if it's beneath a OR because
```

Comment:
missing a star here? (/** and a new line)

Reply:
Intellij has warning if `/**` is used because they think such comments shouldn't be used *inside* of methods. I've no strong opinion on this - can also change it.